feat: QC metrics extraction and validation

ivadym opened the pull request:
### This PR:

Dumps the summarised metrics into a JSON file after performing a metric validation.

Added:

Pydantic models that shape the QC metrics JSON generation and validation.
qc_metrics_summary.json is saved in the analysis output directory.

Changed:

Updated previous workflow regarding delivery & report according to the new QC metrics function.

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

ivadym commented:
Output file saved in analysis/qc:

<img width="435" alt="Screen Shot 2021-09-23 at 16 53 54" src="https://user-images.githubusercontent.com/26309194/134530836-9cbb0805-1519-40be-9a21-1e7edeaaee72.png">

Balsamic version 7.2.6

hassanfa commented on BALSAMIC/utils/models.py:
```diff
@@ -688,3 +691,77 @@ class BalsamicWorkflowConfig(BaseModel):
+    def check_metric(cls, value):
+        """Checks if each metric meets the filtering conditions"""
+        for metric in value:
+            if metric.condition is not None and not eval(
+                f"operator.{metric.condition.norm}({metric.value},{metric.condition.threshold})"
+            ):
```
Use of `eval` almost always makes debugging troublesome.

condition is extracted, then the operator is passed to `operator` to evaluate the metric.

What you can do is to use following:

![image](https://user-images.githubusercontent.com/1086877/134899453-79114610-ef84-45e4-af45-1d87a9594e74.png)

```python
import operator
valid_ops={
"eq": operator.eq,
"lt": operator.lt,
}

fold80_threshold=1.8;
actual_fold80=1.6;

metric_conditon="eq"
valid_ops[metric_conditon](actual_fold80, fold80_threshold)

new_metric_condition="lt"
valid_ops[new_metric_condition](actual_fold80, fold80_threshold)
```


Also, i suggest to `continue` if `metric.conditon` is `None`:

```suggestion
            if metric.condition is None:
                 continue
```

hassanfa commented on BALSAMIC/workflows/balsamic.smk:
```diff
@@ -302,17 +302,33 @@ rule all:
+                jsonFile.write(qc_metrics_summary)
+        except ValueError as val_exc:
+            LOG.error(val_exc)
+            raise BalsamicError("QC metrics do not meet the filtering criteria")
```
As we discussed this morning, this part can be removed for this PR.

hassanfa replied:
And we'd like to log all the failed ones as well as passed ones.

Metrics:
- [❌ ] Failed FOLD 80
- [✔️ ] Passed cov_500
- etc

hassanfa reviewed (changes requested):
I really liked the part with models. Lots of improvements and beautiful code.

Tests are well organized and arranged as well. Although they are testing for multiple items per test function. Not that important.

I have a request for change in 'rule all' and 'eval' part of models. Well written code overall.

ivadym commented:
Thanks for your comments @hassanfa. I've changed the validation workflow in order to retrieve and show both the failed and passed metrics (removing the validation error raise and treatment for now).

Updated output json file (test data with 1.0 as threshold for the fold-80 metric):

>     "S1_R": {
>         "failed": {
>             "FOLD_80_BASE_PENALTY": 1.257243
>         },
>         "passed": {
>             "MEAN_INSERT_SIZE": 225.04495,
>             "MEAN_TARGET_COVERAGE": 1507.434421,
>             "MEDIAN_TARGET_COVERAGE": 1451.0,
>             "PCT_OFF_BAIT": 0.13528,
>             "PCT_TARGET_BASES_1000X": 0.915135,
>             "PCT_TARGET_BASES_100X": 1.0,
>             "PCT_TARGET_BASES_250X": 1.0,
>             "PCT_TARGET_BASES_500X": 0.999134,
>             "PCT_TARGET_BASES_50X": 1.0,
>             "PERCENT_DUPLICATION": 0.102727
>         }
>     },
>     "S2_R": {
>         "failed": {
>             "FOLD_80_BASE_PENALTY": 1.27266
>         },
>         "passed": {
>             "MEAN_INSERT_SIZE": 225.377329,
>             "MEAN_TARGET_COVERAGE": 1510.647921,
>             "MEDIAN_TARGET_COVERAGE": 1456.0,
>             "PCT_OFF_BAIT": 0.134864,
>             "PCT_TARGET_BASES_1000X": 0.917809,
>             "PCT_TARGET_BASES_100X": 1.0,
>             "PCT_TARGET_BASES_250X": 1.0,
>             "PCT_TARGET_BASES_500X": 0.998983,
>             "PCT_TARGET_BASES_50X": 1.0,
>             "PERCENT_DUPLICATION": 0.103587
>         }
>     }

hassanfa reviewed (approved):
Nicely done! CNVKit container is failing, but it is fixed in #780. So you can merge this without worry.

hassanfa commented:
ping @mayabrandi
